build.py

```diff
@@ -343,6 +349,69 @@ def _get_csharp_properties(args: argparse.Namespace):
     return props
 
 
+def _run_android_tests(args, ):
+    # only run the tests on the emulator for x86_64 currently.
+    # TODO: may also be possible to run on a Mac with an arm64 chip
+    if args.android_abi != "x86_64":
+        log.info("Skipping Android tests as they are only supported on x86_64 currently.")
+        return
+
+    if not args.build_java:
+        # currently we only have an Android test app that we run on the emulator to test the Java bindings.
+        log.warning("Android testing requires --build_java to be set.")
+        return
+
+    sdk_tool_paths = util.android.get_sdk_tool_paths(args.android_home)
+    adb = sdk_tool_paths.adb
+    with contextlib.ExitStack() as context_stack:
+        # use API 27 or higher so the emulator is Android 8.1 (2017) or later
+        android_api = max(args.android_api, 27)
+
+        if args.android_run_emulator:
+            avd_name = "ort_genai_android"
+            system_image = f"system-images;android-{android_api};default;{args.android_abi}"
+
+            util.android.create_virtual_device(sdk_tool_paths, system_image, avd_name)
+            emulator_proc = context_stack.enter_context(
+                util.android.start_emulator(
+                    sdk_tool_paths=sdk_tool_paths,
+                    avd_name=avd_name,
+                    extra_args=["-partition-size", "2047", "-wipe-data"],
+                )
+            )
+            context_stack.callback(util.android.stop_emulator, emulator_proc)
+
+        # use the gradle wrapper under <repo root>/java to run the test app on the emulator.
+        # the test app loads and runs a test model using the GenAI Java bindings
+        gradle_executable = str(REPO_ROOT / "src" / "java" / ("gradlew.bat" if util.is_windows() else "gradlew"))
+        android_test_path = args.build_dir / "src" / "java" / "androidtest"
+        import subprocess
+        exception = None
+        try:
+            util.run([gradle_executable, "--no-daemon",
+                      f"-DminSdkVer={android_api}",
+                      "clean",
+                      "connectedDebugAndroidTest"],
+                     cwd=android_test_path,
+                     capture_stdout=True,
+                     capture_stderr=True,)
+        except subprocess.CalledProcessError as e:
+            exception = e
+            print(e)
+            print(f"Output:\n{e.output.decode('utf-8')}")
+            print(f"stderr:\n{e.stderr.decode('utf-8')}")
+
+        # Print test log output so we can easily check that the test ran as expected
+        util.run([adb, "logcat", "-s", "-d", "GenAI:*"])
+        util.run([adb, "logcat", "-s", "-d", "ORTGenAIAndroidTest:*"])
+        util.run([adb, "logcat", "-s", "-d", "TestRunner:*"])
```

Comment:
how about using a single logcat call so the relative ordering of the logs from different tags is shown? could use a filter like `"GenAI:V ORTGenAIAndroidTest:V TestRunner:V *:S"`

Reply:
was this one addressed?

Reply:
Sorry - I tested it and validated it worked but it ended up in a different branch (juggling too many changes at once). I'm going to checkin this PR so it's done, and include the change in the PR to add the Android packaging pipeline.